transport/client.go

```diff
@@ -0,0 +1,310 @@
+}
+
+// RefreshKeys will make sure the Transport has loaded keys and has a
+// valid certificate. It will handle any persistence, check that the
+// certificate is valid (i.e. that its expiry date is within the
+// Before date), and handle certificate reissuance as needed.
+func (tr *Transport) RefreshKeys() (err error) {
+	if !tr.Provider.Ready() {
+		log.Debug("key and certificate aren't ready, loading")
+		err = tr.Provider.Load()
+		if err != nil && err != kp.ErrCertificateUnavailable {
+			log.Debugf("failed to load keypair: %v", err)
+			kr := tr.Identity.Request.KeyRequest
+			if kr == nil {
+				kr = csr.NewBasicKeyRequest()
+			}
+
+			err = tr.Provider.Generate(kr.Algo(), kr.Size())
+			if err != nil {
+				log.Debugf("failed to generate key: %v", err)
+				return
+			}
+		}
+	}
+
+	lifespan := tr.Lifespan()
+	if lifespan < tr.Before {
+		log.Debugf("transport's certificate is out of date (lifespan %s)", lifespan)
+		req, err := tr.Provider.CertificateRequest(tr.Identity.Request)
+		if err != nil {
+			log.Debugf("couldn't get a CSR: %v", err)
+			return err
+		}
+
+		log.Debug("requesting certificate from CA")
+		cert, err := tr.CA.SignCSR(req)
+		if err != nil {
+			log.Debugf("failed to get the certificate signed: %v", err)
+			return err
+		}
+
+		log.Debug("giving the certificate to the provider")
+		err = tr.Provider.SetCertificatePEM(cert)
+		if err != nil {
+			log.Debugf("failed to set the provider's certificate: %v", err)
+			return err
+		}
+
+		log.Debug("storing the certificate")
+		err = tr.Provider.Store()
+		if err != nil {
+			log.Debugf("the provider failed to store the certificate: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (tr *Transport) getCertificate() (cert tls.Certificate, err error) {
+	if !tr.Provider.Ready() {
+		log.Debug("transport isn't ready; attempting to refresh keypair")
+		err = tr.RefreshKeys()
+		if err != nil {
+			log.Debugf("transport couldn't get a certificate: %v", err)
+			return
+		}
+	}
+
+	cert, err = tr.Provider.X509KeyPair()
+	if err != nil {
+		log.Debugf("couldn't generate an X.509 keypair: %v", err)
+	}
+
+	return
+}
+
+// Dial initiates a TLS connection to an outbound server. It returns a
+// TLS connection to the server.
+func Dial(address string, tr *Transport) (*tls.Conn, error) {
+	host, _, err := net.SplitHostPort(address)
+	if err != nil {
+		// Assume address is a hostname, and that it should
+		// use the HTTPS port number.
+		host = address
+		address = net.JoinHostPort(address, "443")
+	}
+
+	cfg, err := tr.TLSClientAuthClientConfig(host)
+	if err != nil {
+		return nil, err
+	}
+
+	return tls.Dial("tcp", address, cfg)
+}
+
+// AutoUpdate will automatically update the listener. If a non-nil
+// certUpdates chan is provided, it will receive timestamps for
+// reissued certificates. If errChan is non-nil, any errors that occur
+// in the updater will be passed along.
+func (tr *Transport) AutoUpdate(certUpdates chan time.Time, errChan chan error) {
```

Comment:
Explicitly use read/write-only chans? Also, do we want to crash the running go process if this goroutine panics?